src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_windows.cc

```diff
@@ -753,131 +743,68 @@ class SockToPolledFdMap {
                            StatusToString(error).c_str());
       return INVALID_SOCKET;
     }
-    GrpcPolledFdWindows* polled_fd =
-        new GrpcPolledFdWindows(s, map->mu_, af, type);
+    auto on_shutdown_locked = [self, s]() {
+      // grpc_winsocket_shutdown calls closesocket which invalidates our
+      // socket -> polled_fd mapping because the socket handle can be henceforth
+      // reused.
+      self->sockets_.erase(s);
```

Comment:
If we are erasing the polled fd from the map, then it's because `ShutdownLocked` was called on the polled fd. This means that the c-ares wrapper actually owns the polled fd - it deletes it [here](https://github.com/grpc/grpc/blob/aaf6c3475cd0de21b96839ccbfb703cf9c798697/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc#L223).

Note that the polled fd factory only deletes those polled fds not still remaining in the map at the end of the resolution - i.e. polled fds created by the c-ares library and not ever seen by the c-ares grpc wrapper.

Reply:
Got it. Thanks for the explanation! I have follow-up questions: Is it because some operations on the socket encountered some errors so that c-ares decided to close it without returning it back to the c-ares wrapper through `ares_getsock()`? In that case, why don't we just actually close that socket in `CloseSocket` and destroy the `GrpcPolledFdWindows` but wait till the end when the factory is destroyed?

Reply:
> Is it because some operations on the socket encountered some errors so that c-ares decided to close it without returning it back to the c-ares wrapper through ares_getsock()?

Yes, this is the practical case I'm thinking of.

> In that case, why don't we just actually close that socket in CloseSocket and destroy the GrpcPolledFdWindows but wait till the end when the factory is destroyed?

We could do this, but I don't see the benefit, and it would add extra complexity because we would need to track additional state per polled fd: something like `bool owned_by_c_ares_lib_` which would be set `true` in `NewGrpcPolledFdLocked`.

`CloseSocket` would then need to check `owned_by_c_ares_lib_`, and if true we'd need to:

1) shut down the endpoint

2) remove the entry from the `sockets_` map

3) delete the endpoint


IMO, that would be more complex than the current model of this PR, where we just shutdown/destroy everything left in `sockets_` at the end.

Reply:
SGTM, I'm all for simplifying this code.